src/dfx/src/lib/deps.rs

```diff
@@ -0,0 +1,264 @@
+use crate::lib::{environment::Environment, error::DfxResult};
+use dfx_core::{
+    config::cache::get_cache_root,
+    fs::composite::ensure_parent_dir_exists,
+    json::{load_json_file, save_json_file},
+};
+
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{anyhow, bail, Context};
+use candid::Principal;
+use fn_error_context::context;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct PulledJson {
+    pub canisters: BTreeMap<Principal, PulledCanister>,
+}
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct PulledCanister {
+    // name of `type: pull` in dfx.json. None if indirect dependency.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+
+    pub dependencies: Vec<Principal>,
+    // the hash on chain
+    // wasm_hash if defined
+    // or get from canister_status
+    pub wasm_hash: String,
+
+    pub init_guide: String,
+    // candid:args
+    pub candid_args: Option<String>,
```

Comment:
```suggestion
    #[serde(skip_serializing_if = "Option::is_none")]
    pub candid_args: Option<String>,
```

Reply:
`candid:args` is a required field. I will turn it into non-optional string.
```rust
pub candid_args: String
```